app_dart/lib/src/service/scheduler/graph.dart

```diff
@@ -2,20 +2,51 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:github/github.dart';
+import 'package:meta/meta.dart';
+import 'package:retry/retry.dart';
 import 'package:yaml/yaml.dart';
 
+import '../../foundation/providers.dart';
+import '../../foundation/utils.dart';
 import '../../model/proto/internal/scheduler.pb.dart';
+import '../config.dart';
 
 /// Load [yamlConfig] to [SchedulerConfig] and validate the dependency graph.
 SchedulerConfig schedulerConfigFromYaml(YamlMap? yamlConfig) {
   final SchedulerConfig config = SchedulerConfig();
   config.mergeFromProto3Json(yamlConfig);
-  _validateSchedulerConfig(config);
+  validateSchedulerConfig(config);
 
   return config;
 }
 
-void _validateSchedulerConfig(SchedulerConfig schedulerConfig) {
+Future<SchedulerConfig> schedulerConfigFromYamlWithBuilderCheck(YamlMap? yamlConfig, {RepositorySlug? slug}) async {
```

Comment:
Thanks for the advice!

Sorry for the github commits, I have squashed them now, but I guess the patches from other contributors still pertain in the overall change lists, which makes it difficult to review. Sorry about that. 

FOR REVIEW:

Yeah I was duplicating schedulerConfigFromYaml because with one declaration of schedulerConfigFromYaml, the process of pulling from the tip of tree makes the function async. And if schedulerConfigFromYaml is async, every function that depend on it become async. I have now removed the duplicate declarations, and instead make all functions that depend on it to be async as well. 

for _validateSchedulerConfig I removed the _ for testing purpose, so that I can supply yaml files in tests, instead of invoking the actual pulling from tip of tree. 

I have changed the optional parameter of slug to be a required positional parameter, because an optional parameter needs [Default values of an optional parameter must be constant] check, and since we cannot have a const constructor for RepositorySlug, I changed it to be positional.